sdk/identity/azure-identity-v2/src/main/java/com/azure/identity/v2/AuthorizationCodeCredential.java

```diff
@@ -0,0 +1,128 @@
+ * and applications. It provides a secure and scalable way to authenticate users and grant them access to Azure
+ * resources.
+ * The AuthorizationCodeCredential authenticates a user or an application and acquires a token with the configured
+ * authorization code and the redirectURL where authorization code was received.</p>
+ *
+ * <p><strong>Sample: Construct AuthorizationCodeCredential</strong></p>
+ *
+ * <p>The following code sample demonstrates the creation of a {@link AuthorizationCodeCredential},
+ * using the {@link AuthorizationCodeCredentialBuilder} to configure it.
+ * The {@code authorizationCode}, {@code redirectUrl} and {@code clientId} are required to be configured to create
+ * {@link AuthorizationCodeCredential}. Once this credential is created, it may be passed into the builder of many of
+ * the Azure SDK for Java client builders as the 'credential' parameter.</p>
+ *
+ * <pre>
+ * TokenCredential authorizationCodeCredential = new AuthorizationCodeCredentialBuilder&#40;&#41;.authorizationCode&#40;
+ *         &quot;&#123;authorization-code-received-at-redirectURL&#125;&quot;&#41;
+ *     .redirectUrl&#40;&quot;&#123;redirectUrl-where-authorization-code-is-received&#125;&quot;&#41;
+ *     .clientId&#40;&quot;&#123;clientId-of-application-being-authenticated&quot;&#41;
+ *     .build&#40;&#41;;
+ * </pre>
+ *
+ * @see com.azure.identity.v2
+ * @see AuthorizationCodeCredentialBuilder
+ */
+public class AuthorizationCodeCredential implements TokenCredential {
+    private static final ClientLogger LOGGER = new ClientLogger(AuthorizationCodeCredential.class);
+
+    private final ConfidentialClient confidentialClient;
+    private final PublicClient publicClient;
+    private final PublicClientOptions publicClientOptions;
+    private final AtomicReference<MsalAuthenticationAccount> cachedToken;
+    private boolean isCaeEnabledRequestCached;
+    private boolean isCaeDisabledRequestCached;
+    private boolean isCachePopulated;
+
+    /**
+     * Creates an AuthorizationCodeCredential with the given identity client options.
+     *
+     * @param clientSecret the client secret of the application
+     * @param publicClientOptions the options for configuring the identity client
+     */
+    AuthorizationCodeCredential(String clientSecret, PublicClientOptions publicClientOptions) {
+        this.publicClient = new PublicClient(publicClientOptions);
+        this.publicClientOptions = publicClientOptions;
+        if (!CoreUtils.isNullOrEmpty(clientSecret)) {
+            confidentialClient = new ConfidentialClient(
+                (ConfidentialClientOptions) new ConfidentialClientOptions().setClientSecret(clientSecret)
+                    .setClientId(publicClientOptions.getClientId())
+                    .setTenantId(publicClientOptions.getTenantId()));
+        } else {
+            confidentialClient = null;
+        }
+        this.cachedToken = new AtomicReference<>();
```

Comment:
I wonder if we can generalize this caching mechanism for the creds that use it. What if we had an `AzureCredential` base class or something that we extend from to get it?

Reply:
Abstracted into a brokered auth cache API.